synapse/metrics/__init__.py

```diff
@@ -182,6 +184,18 @@ def f(*args, **kwargs):
         end = time.time() * 1000
         tick_time.inc_by(end - start)
         pending_calls_metric.inc_by(num_pending)
+
+        threshold = gc.get_threshold()
+        counts = gc.get_count()
+
+        start = time.time() * 1000
+        for i in [2, 1, 0]:
+            if threshold[i] < counts[i]:
+                logger.info("Collecting gc %d", i)
+                gc.collect(i)
+        end = time.time() * 1000
+        gc_time.inc_by(end - start)
```

Comment:
Might be interesting to collect stats about the different levels of collection going on.